### PR TITLE
Add is_finalized

### LINKED ISSUE #5208: Consider changing the behavior of `Kokkos::is_initialized` and providing `Kokkos::is_finalized`?
Currently `Kokkos::is_initialized` returns
* `false` before calling `Kokkos::initialize` or after calling `Kokkos::finalize`, and
* `true` in-between.

This is unfortunate since it is not legal to call `Kokkos::initialize` more than once, and it means there is no way for programs to figure out if they can safely do so.

We should consider
* changing `Kokkos::is_initialized` so that it returns `false` before calling `Kokkos::initialize`, and `true` after,
* adding `Kokkos::is_finalized` that would return `false` before calling `Kokkos::finalize`, and `true` after.

This would be a breaking change so we need to think carefully about it.

```C++
#include <Kokkos_Core.hpp>

int main(int argc, char* argv[]) {
  assert(!Kokkos::is_initialized());
  assert(!Kokkos::is_finalized());

  Kokkos::initialize(argc, argv);

  assert(Kokkos::is_initialized());
  assert(!Kokkos::is_finalized());

  Kokkos::finalize(argc, argv);

  assert(Kokkos::is_initialized());  // <- breaking change
  assert(Kokkos::is_finalized());
```

### PR DESCRIPTION
**Pull request by crtrott**
Fixes #5208.

is_finalized is also checked in the ScopeGuard constructor

Co-Authored-By: Damien L-G <dalg24@gmail.com>

### CODE REVIEW

File: core/src/Kokkos_Core.hpp [hunk 1]
```diff
@@ -202,7 +203,12 @@ class KOKKOS_ATTRIBUTE_NODISCARD ScopeGuard {
 #endif
   ScopeGuard(int& argc, char* argv[]) {
     sg_init = false;
-    if (!Kokkos::is_initialized()) {
+    if (!is_initialized()) {
+      if (is_finalized()) {
+        Kokkos::abort(
+            "Attempt to create a Kokkos::ScopeGuard after Kokkos::finalize was "
+            "called.");
+      }
```
Comment (dalg24) [inline core/src/Kokkos_Core.hpp:211]:
```suggestion
    if (is_finalized()) {
      Kokkos::abort(
          "Attempt to create a Kokkos::ScopeGuard after Kokkos::finalize was "
          "called.");
    }
    if (!is_initialized()) {
```

File: core/src/Kokkos_Core.hpp [hunk 2]
```diff
@@ -214,14 +220,19 @@ class KOKKOS_ATTRIBUTE_NODISCARD ScopeGuard {
   ScopeGuard(
       const InitializationSettings& settings = InitializationSettings()) {
     sg_init = false;
-    if (!Kokkos::is_initialized()) {
+    if (!is_initialized()) {
+      if (is_finalized()) {
+        Kokkos::abort(
+            "Attempt to create a Kokkos::ScopeGuard after Kokkos::finalize was "
+            "called.");
+      }
```
Comment (dalg24) [inline core/src/Kokkos_Core.hpp:228]:
```suggestion
    if (is_finalized()) {
      Kokkos::abort(
          "Attempt to create a Kokkos::ScopeGuard after Kokkos::finalize was "
          "called.");
    }
    if (!is_initialized()) {
```

File: core/src/Kokkos_Core.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Kokkos_Core.hpp:211]: it is one more check in the case of `is_initialized()==true` but I guess thats not that much cost?

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Cuda timing based test failing in one of the CUDA builds

**Review by masterleinad — APPROVED** [review]

**Review by dalg24 — CHANGES_REQUESTED** [review]
On 2nd thought I do not agree with the `ScopeGuard` changes.  Please revert them.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Failure (machine running out of space) are unrelated

**Pull request merged**